packages/flutter/lib/src/widgets/text_selection.dart

```diff
@@ -824,9 +825,41 @@ class _TextSelectionHandleOverlayState
         break;
     }
 
+    // On some platforms we may want to calculate the start and end handles
+    // separately so they scale for the selected content.
+    //
+    // For the start handle we compute the rectangles that encompass the range
+    // of the first full selected grapheme cluster at the beginning of the selection.
+    //
+    // For the end handle we compute the rectangles that encompass the range
+    // of the last full selected grapheme cluster at the end of the selection.
+    final InlineSpan span = widget.renderObject.text!;
+    final String text = span.toPlainText();
+    final int firstSelectedGraphemeExtent;
+    final int lastSelectedGraphemeExtent;
+    if(text.isEmpty || widget.selection.isCollapsed){
+      // The call to selectedGraphemes.characters.first/last will throw a state
+      // error if the given string is empty, so fall back to first/last character
+      // range in this case.
+      //
+      // The call to widget.selection.textInside(text) will return a RangeError
+      // for a collapsed selection, fall back to this case when that happens.
+      firstSelectedGraphemeExtent = 1;
+      lastSelectedGraphemeExtent = 1;
+    }else{
+      final String selectedGraphemes = widget.selection.textInside(text);
+      firstSelectedGraphemeExtent = selectedGraphemes.characters.first.length;
+      lastSelectedGraphemeExtent = selectedGraphemes.characters.last.length;
+    }
+
+    final Rect? startHandleRect = widget.renderObject.getRectForComposingRange(TextRange(start: widget.selection.start, end: widget.selection.start + firstSelectedGraphemeExtent));
```

Comment:
nit: Could you add an assert here that makes sure the `Rect` isn't null when the grapheme extent is greater than 0?

Reply:
When the selection only contains a trailing/leading new line character might be another interesting edge case to check, as the returned box may be as high as 2 lines.